vod/hls/m3u8_builder.c

```diff
@@ -511,48 +511,56 @@ m3u8_builder_build_master_playlist(
 	media_set_t* media_set,
 	vod_str_t* result)
 {
-	media_track_t** cur_sequence_tracks;
-	media_sequence_t* cur_sequence;
-	media_track_t* track;
+	media_sequence_t* cur_sequence;	
 	media_info_t* video;
 	media_info_t* audio = NULL;
 	uint32_t sequence_index;
 	uint32_t bitrate;
 	u_char* p;
-	size_t max_video_stream_inf;
-	size_t result_size;
+	size_t          max_video_stream_inf = 0;
+	size_t          result_size = 0;
+        uint32_t        main_media_type;
+        media_track_t*  cur_track;
+	media_track_t* cur_sequence_tracks[MEDIA_TYPE_COUNT + 1];
 
-	// calculate the result size
-	max_video_stream_inf = 
-		sizeof(m3u8_stream_inf_video) - 1 + 3 * VOD_INT32_LEN + MAX_CODEC_NAME_SIZE +
-		MAX_CODEC_NAME_SIZE + 1 +
-		sizeof(m3u8_stream_inf_suffix) - 1;
-	result_size = 
-		sizeof(m3u8_header) + 
-		media_set->sequence_count * max_video_stream_inf;		// using only video since it's larger than audio
 
-	for (cur_sequence = media_set->sequences; cur_sequence < media_set->sequences_end; cur_sequence++)
-	{
-		cur_sequence_tracks = cur_sequence->filtered_clips[0].longest_track;
 
-		track = (cur_sequence_tracks[MEDIA_TYPE_VIDEO] != NULL ?
-			cur_sequence_tracks[MEDIA_TYPE_VIDEO] :
-			cur_sequence_tracks[MEDIA_TYPE_AUDIO]);
-		if (base_url->len != 0)
-		{
-			result_size += base_url->len + 1;
-			if (track->file_info.uri.len > 0)
+	// calculate the result size
+	result_size = sizeof(m3u8_header);
+	for (cur_sequence = media_set->sequences; cur_sequence < media_set->sequences_end; cur_sequence++)
+	{	    
+		main_media_type = cur_sequence->track_count[MEDIA_TYPE_VIDEO] != 0 ? MEDIA_TYPE_VIDEO : MEDIA_TYPE_AUDIO;
+             
+		cur_sequence_tracks[MEDIA_TYPE_VIDEO] = NULL;
+		cur_sequence_tracks[MEDIA_TYPE_AUDIO] = cur_sequence->filtered_clips[0].longest_track[MEDIA_TYPE_AUDIO];
+		for (cur_track = cur_sequence->filtered_clips[0].first_track; cur_track < cur_sequence->filtered_clips[0].last_track; cur_track++)
+		{			
+			if (cur_track->media_info.media_type != main_media_type)
 			{
-				result_size += track->file_info.uri.len;
+				continue;
 			}
-			else
+			max_video_stream_inf += 
+				sizeof(m3u8_stream_inf_video) - 1 + 3 * VOD_INT32_LEN + MAX_CODEC_NAME_SIZE +
+				MAX_CODEC_NAME_SIZE + 1 +
+				sizeof(m3u8_stream_inf_suffix) - 1;
```

Comment:
this calculation is constant please move it back to the beginning of the function